src/robot/variables/replacer.py

```diff
@@ -137,46 +137,59 @@ def _get_variable_item(self, match, value):
             logger.warn("Accessing variable items using '%s' syntax "
                         "is deprecated. Use '$%s' instead." % (var, var[1:]))
         for item in match.items:
-            if is_dict_like(value):
-                value = self._get_dict_variable_item(name, value, item)
-            elif is_list_like(value):
-                value = self._get_list_variable_item(name, value, item)
+            if is_sequence(value):
+                value = self._get_sequence_variable_item(name, value, item)
+            elif is_subscriptable(value):
+                value = self._get_subscriptable_variable_item(name, value, item)
             else:
                 raise VariableError(
-                    "Variable '%s' is %s, not list or dictionary, and thus "
-                    "accessing item '%s' from it is not possible."
-                    % (name, type_name(value), item)
+                    "Variable '%s' is %s, which is not subscriptable, and "
+                    "thus accessing item '%s' from it is not possible. To use "
+                    "'[%s]' as a literal value, it needs to be escaped like "
+                    "'\\[%s]'." % (name, type_name(value), item, item, item)
                 )
             name = '%s[%s]' % (name, item)
         return value
 
-    def _get_list_variable_item(self, name, variable, index):
+    def _get_sequence_variable_item(self, name, variable, index):
         index = self.replace_string(index)
         try:
-            index = self._parse_list_variable_index(index, name[0] == '$')
+            index = self._parse_sequence_variable_index(index, name[0] == '$')
         except ValueError:
-            raise VariableError("List '%s' used with invalid index '%s'."
-                                % (name, index))
+            raise VariableError("Sequence '%s' used with invalid index '%s'. "
```

Comment:
"Sequence" is better than "List" in this context but perhaps `type_name(variable)` would be even better. In here we needed to capitalize the returned word, though, and should take into account that e.g. `"MyClass".capitalize()` yields `"Myclass"` which isn't what we want. Probably would be best to add `capitalize=False` argument to `type_name()` to handle this stuff there, but I'm not certain should it be done as part of this PR or should it be done afterwards.

Reply:
Hmm, perhaps I should enhance `type_name()` in this regard separately. It could then be used here after merging changes from master.

Reply:
Done in d7110621c16a371e72b541b70f01cbea7f0f4c8e